xds/internal/balancer/edsbalancer/balancergroup.go

```diff
@@ -233,44 +260,105 @@ func (bg *balancerGroup) add(id internal.Locality, weight uint32, builder balanc
-	if _, ok := bg.idToBalancerConfig[id]; ok {
-		bg.outgoingMu.Unlock()
-		grpclog.Warningf("balancer group: adding a balancer with existing ID: %s", id)
-		return
+	var sbc *subBalancerWithConfig
+	// If outgoingStarted is true, search in the cache. Otherwise, cache is
+	// guaranteed to be empty, searching is unnecessary.
+	if bg.outgoingStarted {
+		if temp, ok := bg.balancerCache.Remove(id); ok {
+			sbc, _ = temp.(*subBalancerWithConfig)
+			if sbc != nil && sbc.builder != builder {
+				// If the sub-balancer in cache was built with a different
+				// balancer builder, don't use it, cleanup this old-balancer,
+				// and behave as sub-balancer is not found in cache.
+				//
+				// NOTE that this will also drop the cached addresses for this
+				// sub-balancer, which seems to be reasonable.
+				delete(bg.idToBalancerConfig, id)
+				sbc.stopBalancer()
+				bg.cleanupSubConns(id)
+				sbc = nil
+			}
+		}
 	}
-	sbc := &subBalancerWithConfig{
-		builder: builder,
+	if sbc == nil {
+		sbc = &subBalancerWithConfig{
+			ClientConn: &balancerGroupCC{
+				ClientConn: bg.cc,
+				id:         id,
+				group:      bg,
+			},
+			builder: builder,
+		}
+		if bg.outgoingStarted {
+			// Only start the balancer if bg is started. Otherwise, we only keep the
+			// static data.
+			sbc.startBalancer()
+		}
+	} else {
+		// When brining back a sub-balancer from cache, re-send the cached
+		// picker and state.
+		sbc.updateBalancerStateWithCachedPicker()
 	}
 	bg.idToBalancerConfig[id] = sbc
-
-	if bg.outgoingStarted {
-		// Only start the balancer if bg is started. Otherwise, we only keep the
-		// static data.
-		sbc.startBalancer(&balancerGroupCC{
-			ClientConn: bg.cc,
-			id:         id,
-			group:      bg,
-		})
-	}
 	bg.outgoingMu.Unlock()
 }
 
-// remove removes the balancer with id from the group, and closes the balancer.
+// remove removes the balancer with id from the group.
+//
+// But doesn't close the balancer. The balancer is kept in a cache, and will be
+// closed after timeout. Cleanup work (closing sub-balancer and removing
+// subconns) will be done after timeout.
 //
 // It also removes the picker generated from this balancer from the picker
 // group. It always results in a picker update.
 func (bg *balancerGroup) remove(id internal.Locality) {
 	bg.outgoingMu.Lock()
 	if bg.outgoingStarted {
-		if config, ok := bg.idToBalancerConfig[id]; ok {
-			config.stopBalancer()
+		if subBalancerToBeRemoved, ok := bg.idToBalancerConfig[id]; ok {
+			// Add sub-balancer to cache only when this id exists.
+			bg.balancerCache.Add(id, subBalancerToBeRemoved, func() {
+				// After timeout, when sub-balancer is removed from cache, need
+				// to also remove it from map, close the underlying
+				// sub-balancer, and remove all its subconns.
+				bg.outgoingMu.Lock()
+				if bg.outgoingStarted {
+					delete(bg.idToBalancerConfig, id)
+					subBalancerToBeRemoved.stopBalancer()
+				}
+				bg.outgoingMu.Unlock()
+				bg.cleanupSubConns(id)
```

Comment:
outgoingMu isn't very important for cleanupSubConns(). It's held in add only because cleanupSubConns() needs to happen first, and there's no way to unlock in between.